### PR TITLE
[luci] Support rank4 in ReplaceMulAddWithDConv

### PR DESCRIPTION
**Pull request by jinevening**
This supports rank4 gamma/beta in ReplaceMulAddWithDConv.

ONE-DCO-1.0-Signed-off-by: Hyukjin Jeong <hj1.jeong@samsung.com>

### CODE REVIEW

File: compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.test.cpp [hunk 1]
```diff
@@ -141,6 +141,37 @@ TEST(ReplaceMulAddWithDepthwiseConv, simple)
   }
 }
 
+TEST(ReplaceMulAddWithDepthwiseConv, simple_rank4)
```
Comment (seanshpark) [inline compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.test.cpp:144]: (optional) would be nice to add negative test with rank 2 or 3.

**Review by seanshpark — APPROVED** [review]
LGTM

File: compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.cpp [hunk 2]
```diff
@@ -26,8 +26,13 @@ namespace
...
 {
-  assert(gamma->rank() == 1);
-  auto channel_size = gamma->dim(0).value();
+  assert(gamma->rank() == 1 or gamma->rank() == 4);
+
+  uint32_t channel_size = 0;
+  if (gamma->rank() == 1)
+    channel_size = gamma->dim(0).value();
+  else if (gamma->rank() == 4)
+    channel_size = gamma->dim(3).value();
```
Comment (mhs4670go) [inline compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.cpp:35]: How about this one?
```suggestion
  uint32_t channel_idx = gamma->rank() - 1;
  uint32_t channel_size = gamma->dim(channel_idx).value();
```

File: compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.cpp [hunk 3]
```diff
@@ -53,8 +58,13 @@ luci::CircleConst *create_weights_from_gamma(luci::CircleConst *gamma)
...
 {
-  assert(beta->rank() == 1);
-  auto channel_size = beta->dim(0).value();
+  assert(beta->rank() == 1 or beta->rank() == 4);
+
+  uint32_t channel_size = 0;
+  if (beta->rank() == 1)
+    channel_size = beta->dim(0).value();
+  else if (beta->rank() == 4)
+    channel_size = beta->dim(3).value();
```
Comment (mhs4670go) [inline compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.cpp:67]: ditto

**Review by mhs4670go — DISMISSED** [review]
LGTM

File: compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.cpp [hunk 4]
```diff
@@ -28,11 +28,10 @@ luci::CircleConst *create_weights_from_gamma(luci::CircleConst *gamma)
...
 
-  uint32_t channel_size = 0;
-  if (gamma->rank() == 1)
-    channel_size = gamma->dim(0).value();
-  else if (gamma->rank() == 4)
-    channel_size = gamma->dim(3).value();
+  uint32_t channel_idx = gamma->rank() - 1;
+  uint32_t channel_size = gamma->dim(channel_idx).value();
+  for (uint32_t i = 0; i < channel_idx; i++)
+    assert(gamma->dim(i).value() == 1); // FIX is_batchnorm_mul UNLESS
```
Comment (seanshpark) [inline compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.cpp:34]: Could you explain the intention of this for loop?

Comment (jinevening) [reply compiler/luci/pass/src/ReplaceMulAddWithDepthwiseConvPass.cpp:34]: The intention is to check whether gamma/beta's shape is broadcastable in the channel direction.

For example, if channel size is 16, gamma/beta's shape can be {16}, {1, 16}, {1, 1, 16}, {1,1,1,16}.

**Review by mhs4670go — APPROVED** [review]
LGTM

**Review by seanshpark — APPROVED** [review]
LGTM thank you!

**Pull request merged**